### ISSUE TITLE
[CRITICAL] Datasource resource leak

### ISSUE THREAD

**Issue by sgdc3** [Xephi/AuthMeReloaded#719]
Critical issue, needs to be fixed before anything else.

Original comments (a bit rude):
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6639
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6642
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6643
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6651
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6654
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6656

Video explaining the issue:
https://www.youtube.com/watch?v=dIh4n3tTIjY&feature=youtu.be

**Comment by sgdc3** [comment]
@Xephi @ljacqu @games647 @DNx5 @timvisee

**Comment by sgdc3** [comment]
Possible causes (please add them here):
- Wrong HikariCP usage
- HikariCP shouldn't be used in AuthMe
- Complex SQL Queries
- Complex SQL schema
- Wrong usage of Asynch tasks
- Something synch that blocks the main thread

**Comment by sgdc3** [comment]
It seems that chat spam is causing tps drop on the main thread (probably the cache isn't working as it should)

**Comment by ljacqu** [comment]
What exactly is the issue?, how does it manifest itself? I don't understand the comment thread.

**Comment by games647** [comment]
My first part where he shows authme is authme 3.X. There was blocking SQL on the main thread.

Because of the high CPU Usage at the end (5.X), my guess is that it's about thread-creation. The threads are waiting on the MySQL instance to response while others will be created using Bukkit's cached pooled executor and the old ones cannot be reused because they are still waiting.

In order to analyze it better he or we should work with VisualVM or Java Mission Control. For better analysis we need:
* Thread dumps
* maybe heap dumps
* or even better Java Flight Recordings

**Comment by sgdc3** [comment]
Could you check my commit? It should avoid crashw with spambot.

**Comment by ljacqu** [comment]
@games647 One thing, not sure if it relates, but are we sure all of those DataSource methods need to be `synchronized`? I feel like this should be something that the database already handles, no?

http://stackoverflow.com/questions/1346174/do-i-need-to-synchronize-if-i-have-multiple-threads-that-will-insert-to-mysql
http://stackoverflow.com/questions/1531073/is-java-sql-connection-thread-safe
http://stackoverflow.com/questions/578904/how-do-synchronized-static-methods-work-in-java

**Comment by games647** [comment]
Good point. No they shouldn't be synchronized databases can handle that really good. 

But they should lock (synchronize) if java accesses critical parts. Like non-thread-safe collections or race conditions. For example in SQL:

Same player, java conditional queries - pseudo code

if (sql.blue)
    set red

separate thread

if (sql.blue)
    set yellow

Then the value could be yellow although the first thread started earlier.

But first this is **very very** rare and second then we don't need to lock the complete class/object if would be enough to lock queries for the specific player only. There is also the possibility to use transactions or explicit query locking for that (lockForUpdate). 

So yes, databases do row locking. Using synchronized in this context would lock a lot of other threads.

**Comment by ljacqu** [comment]
@games647 What do you suggest? Should we try to remove `synchronized` on the data source methods and run a stress test?

**Comment by sgdc3** [comment]
done on the 719 bransh

**Comment by games647** [comment]
It would really great if we have data like Java Flight Recordings. We would have info about where are the hot methods, how many threads are waiting (and where), and ...

**Comment by sgdc3** [comment]
There was an issue with HikaryCP, the pool size was dinamic instead of static, that was causing issue with multiple access to the MySql DB.

**Comment by ljacqu** [comment]
@sgdc3 Thanks!

Can't say I'm very happy about 4d06b63 though, looking at the diff doesn't help to understand the changes because existing code was just commented out and new, similar code added. So now one has to scroll up and down to see the changes. Nice touch on using `event.getRecipients()` though! Just a shame it took me five minutes to understand that change :)...

**Comment by sgdc3** [comment]
@ljacqu lol, you know, from mobile it's hard to edit massive amount of code xD

**Comment by sgdc3** [comment]
Update:
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6667
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6668
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6670

**Comment by sgdc3** [comment]
Poor console thread! 
http://dev.bukkit.org/bukkit-plugins/authme-reloaded/?comment=6672

**Comment by sgdc3** [comment]
I think that `dataSource.isAuthAvailable` is the main cause.

**Comment by sgdc3** [comment]
I guess what happens with `DataSource.caching: false`

**Comment by Sitmobedaf** [comment]
1034.
Server with AuthMe: http://pastebin.com/hfirBAPC
During the attack, there was a strong pressure on the hard drive.

Without AuthMe: https://timings.spigotmc.org/?url=16567682

**Comment by sgdc3** [comment]
@Sitmobedaf Mysql or Sqlite?

**Comment by sgdc3** [comment]
What's the value of the 
DataSource.caching
setting? @Sitmobedaf

**Comment by Sitmobedaf** [comment]
sgdc3?, caching: true. SQLite
It does not matter, SQLite or MySQL, they work the same way, if everything is configured correctly and have enough resources.
I'm on a notebook do not want to install the MySQL-server. While I see no reason, the problem lies elsewhere.
I now check with caching: false
10-20 minutes...

**Comment by Sitmobedaf** [comment]
catching:false - all the same.
Hard disk load is 100% server crash just as quickly.

**Comment by ljacqu** [comment]
@Sitmobedaf This is without Antibot, right? While you've discovered an important issue regarding our performance, it's important not to forget that we offer an antibot feature.

**Comment by sgdc3** [comment]
@ljacqu Yeah, but if we could do something better without it we should enhance it. ;)

**Comment by sgdc3** [comment]
The main issue is our caching system, it was an issue in authme since 3.X, now we need to fix it

**Comment by sgdc3** [comment]
The antibot should be an additional security layer, but the server should work fine also without crashing.

**Comment by sgdc3** [comment]
@games647 did you understand something about LogIt and how it handles the database caching?

**Comment by sgdc3** [comment]
Somehow we do more queries during the login process, we need to know the reason.

**Comment by sgdc3** [comment]
@Sitmobedaf could you join our chatroom on gitter? ;) https://gitter.im/Xephi/AuthMeReloaded

**Comment by Sitmobedaf** [comment]
I'm going to try now...

**Comment by games647** [comment]
Could you make a "Java-Flight-Recording"? If you don't know how here is a step by step guide.

1. Start the server with authme using MySQL (one test with caching enabled and one without it)
2. Start Java Mission Control. It's located in here: JAVA_INSTALL_DIR/bin/jmc.exe
3. Right click on the server java process and click start flight recording
4. Select 2 Minutes and profiling for event settings
5. Enable the checkbox heap statistics
6. Wait 30 seconds and then start the bot attack
7. Upload the .jfr-file to this ticket.

**Comment by ljacqu** [comment]
@sgdc3 I don't know if I tested wrong, but the "hide chat" function didn't work for me anymore

**Comment by sgdc3** [comment]
uh, really?

**Comment by sgdc3** [comment]
I'll check later

**Comment by sgdc3** [comment]
@ljacqu i'm an idiot! I used a if instead of a while xD Thanks for the tix!

**Comment by sgdc3** [comment]
Does it work with less players per second? @Sitmobedaf

**Comment by ljacqu** [comment]
@sgdc3 np, but I still think it doesn't work. I tested after my commit.

**Comment by sgdc3** [comment]
@Xephi Should we merge the 719 branch?

**Comment by ljacqu** [comment]
We've done some optimizations within this story -> now closing for inactivity. If something is still missing please open a followup issue with a concise goal / problem description.
